### ISSUE TITLE
[feat] `no-unnecessary-optional-prop` rule

### ISSUE THREAD

**Issue by reteps** [Rel1cx/eslint-react#1599]
### Problem Description

Unnecessarily optional props weaken type safety — the component body must handle `undefined` for a value that is never actually omitted. This leads to unnecessary nullish checks, default values that are never used, and a misleading API surface. Over time as codebases evolve, props that were once optional become effectively required, but the type definition isn't updated.

```tsx
function Button({ label, onClick }: { label: string; onClick?: () => void }) {
  //                                                   ^ optional but always provided
}

// Every usage in the project provides onClick:
<Button label="Save" onClick={handleSave} />
<Button label="Cancel" onClick={handleCancel} />
<Button label="Delete" onClick={handleDelete} />
```

This is similar in spirit to `no-unused-props` (defined but never used) — this would catch the inverse-adjacent case (optional but always provided).

### Alternative Solutions

here is no existing ESLint rule (in typescript-eslint, eslint-plugin-react, or this project) that detects this pattern.

## Rule Details

This rule flags React component props that are marked as optional (`?`) but are provided at every JSX call site in the project, suggesting they should be made required.

This requires whole-program analysis across files.

### Invalid

```tsx
// ❌ Bad: onClick is optional but always provided at every call site
function Button({ label, onClick }: { label: string; onClick?: () => void }) {
  return <button onClick={onClick}>{label}</button>;
}

<Button label="Save" onClick={handleSave} />
<Button label="Cancel" onClick={handleCancel} />
```

### Valid

```tsx
// ✅ Good: onClick is required
function Button({ label, onClick }: { label: string; onClick: () => void }) {
  return <button onClick={onClick}>{label}</button>;
}
```

```tsx
// ✅ Good: onClick is optional and genuinely omitted at some call sites
function Button({ label, onClick }: { label: string; onClick?: () => void }) {
  return <button onClick={onClick}>{label}</button>;
}

<Button label="Save" onClick={handleSave} />
<Button label="Display only" />
```

### What about spread props?

When a call site uses spread props (`<Button {...props} />`), it's ambiguous whether the optional prop is included. The rule should skip components where any call site uses spread that could include the prop.

### What about exported components?

Components exported from a package's public API may have external consumers that rely on the prop being optional. The rule could support an option to exclude exported components.

### What about rarely-used components?

A minimum call-site threshold option (e.g., only flag if the component is used N+ times) could reduce false positives on rarely-used components.

### Evaluation Checklist

- [x] I have had problems with the pattern I want to forbid
- [x] I could not find a way to solve the problem by changing the API of the problematic code or introducing a new API
- [x] I have thought very hard about what the corner cases could be and what kinds of patterns this would forbid that are actually okay, and they are acceptable
- [x] I think the rule explains well enough how to solve the issue to make sure beginners are not blocked by it
- [ ] I have discussed this rule with team members, and they all find it valuable
